### PR TITLE
Symlink support for imfile

### PR DESCRIPTION
**Pull request by jvymazal**
this introduces symlink detection and following as well
as monitoring changes on them.

### CODE REVIEW

**Comment by jvymazal** [comment]
This is attempt to address #2229

File: plugins/imfile/imfile.c [hunk 1]
```diff
@@ -543,6 +546,81 @@ wdmapDel(const int wd)
...
+{
+	const char fullpath[MAXFNAME];
+	const char *ptr = fullpath;
+	struct stat fileInfo;
+	int is_symlink = 0;
+
+	if (file[0] == '/') {
+		ptr = file;
+	} else {
+		snprintf((char *)fullpath, MAXFNAME, "%s/%s", dir, file);
```
Comment (richm) [inline plugins/imfile/imfile.c]:
```
fullpath[MAXFNAME-1] = '\0';
```
to ensure null termination

File: plugins/imfile/imfile.c [hunk 2]
```diff
@@ -543,6 +546,81 @@ wdmapDel(const int wd)
...
+	}
+	if(lstat(ptr, &fileInfo) != 0) {
+		perror("lstat");
```
Comment (richm) [inline plugins/imfile/imfile.c]: use rs_strerror_r here, as used below

Comment (richm) [reply plugins/imfile/imfile.c]: and - should return is_symlink = -1 here too?

File: plugins/imfile/imfile.c (hunk 1 above)
Comment (rgerhards) [reply plugins/imfile/imfile.c]: @richm I had a discussion today on the need to do this sanity check. I also think it's the right thing to do, but I could not find any evidence we are right. Do you have evidence?

File: tests/imfile-symlink.sh [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+rsyslogconf_bak=`mktemp`
+cp $rsyslogconf $rsyslogconf_bak
+watchdir0="/var/log/test0"
+targetdir0="/var/log/target0"
+reltarget0="../target0"
+watchdir1="/var/log/test1"
+targetdir1="/var/log/target1"
+reltarget1="../target1"
+
+if [ "$EUID" -ne 0 ]; then
```
Comment (rgerhards) [inline tests/imfile-symlink.sh]: do we have a chance to get a test without root permissions?

File: plugins/imfile/imfile.c (hunk 1 above)
Comment (richm) [reply plugins/imfile/imfile.c]: @rgerhards I've always had to ensure proper null termination of strings when using snprintf.  I don't have any evidence, just many years of doing it this way.

Comment (rgerhards) [reply plugins/imfile/imfile.c]: @richm I feared that answer... much like I knew and many do. But may look like it is no longer true. Should you run over some case where it does not work, let me know ;-)

Comment (richm) [reply plugins/imfile/imfile.c]: > may look like it is no longer true.

How so?  Was snprintf changed to always do null termination?  To my knowledge (man snprintf on Fedora 27) it was no.  Was there some change to the compiler to ensure `const char fullpath[MAXFNAME]` is initialized to all nulls?  again, not to my knowledge.

Comment (rgerhards) [reply plugins/imfile/imfile.c]: @richm actually yes, snprintf is specified to do that in C99 (here a link to POSIX): http://pubs.opengroup.org/onlinepubs/007904875/functions/fprintf.html

It is interesting that the man page actually does not make a clear guarantee in that regard. I checked the Fedora 27 man pages, as well as Ubuntu and BSD, and indeed they do not specify it precisely - but they claim C99 compatibility. Web searches often bring up discussions where it is said this could be treated as granted.

Red Hat's Florian Weimer wrote a longer post about the problems with ```snprintf```: https://access.redhat.com/blogs/766093/posts/1976193 In it, he states
```
This code assumes that the snprintf implementation does not write an 
unterminated string to the destination buffer on error, which is quite
reasonable as far as such assumptions go.  
```
Given his obviously in-depth resarch, I would guess that he would have noticed if there is a problem.

In conclusion, to me it looks like this is really something that we can rely on - but I am still puzzled. As I said, I was also thinking that there may be a problem on some platforms, but I cannot find hard evidence of where these problems really exist. Having pointers would be most useful.

Comment (rgerhards) [reply plugins/imfile/imfile.c]: I should have also provided this quote from POSIX:

*The snprintf() function shall be equivalent to sprintf(), with the addition of the n argument which states the size of the buffer referred to by s. If n is zero, nothing shall be written and s may be a null pointer. Otherwise, output bytes beyond the n-1st shall be discarded instead of being written to the array, and a null byte is written at the end of the bytes actually written into the array.*

It's pretty much at the top of the document.

Comment (richm) [reply plugins/imfile/imfile.c]: Ok.  That seems to be the behavior on fedora 27 (gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)) and rhel7 (gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)).  Trying various ways to break snprintf with a simple test program shows that it always null terminates the string and returns the correct result code.

Consider my objection withdrawn.

File: tests/imfile-symlink.sh (hunk 3 above)
Comment (jvymazal) [reply tests/imfile-symlink.sh]: Well, the test needs permissions to restart rsyslog to properly flush the watches while re-creating test dirs, so not likely. Is it a big problem when we have root test?

Comment (richm) [reply tests/imfile-symlink.sh]: Yes.  Why not write a test that uses the test framework, which is able to configure and run rsyslog with out redhat/fedora-isms such as running as root, running using systemd, etc.?
For example: https://github.com/rsyslog/rsyslog/blob/master/tests/imfile-wildcards-dirs-multi.sh

Comment (rgerhards) [reply tests/imfile-symlink.sh]: +1 - we should have tests for this functionality. If it is just a start, that's "biz as usual" inside the testbench.

Comment (jvymazal) [reply tests/imfile-symlink.sh]: Yes, I am planning to rewrite the test, what is a timeframe for this to make the 8.36 please?

**Comment by rgerhards** [comment]
8.36 officially was closed friday, I am doing final merges at the moment. So we are at 8.37. BTW: the timeframe is always shown in the assigned milestone, release dates on https://www.rsyslog.com.

File: plugins/imfile/imfile.c [hunk 4]
```diff
@@ -770,12 +853,91 @@ poll_tree(fs_edge_t *const chld)
...
+				}
+				if(rv == 0) { /* out of memory */
+					LogError(errno, RS_RET_OUT_OF_MEMORY, "imfile: poll_tree: \
+							hastable_insert failed");
+					if (freetarget) {
+						free(target);
+					}
+					continue;
+				}
+				DBGPRINTF("imfile: %s -> %s added to symlink hash\n", slink, target);
```
Comment (nhosoi) [inline plugins/imfile/imfile.c]: Do we need to call realloc to expand files.gl_pathv to allocate the space for target prior to the line 931 (as being done in the downstream version)?

Comment (jvymazal) [reply plugins/imfile/imfile.c]: Yes, I will edit the PR once I finish first round of testing.

**Comment by jvymazal** [comment]
@rgerhards I have two questions now. First, how can I determine which part of my test causes the "permission denied" failures (as far as I am aware I am not doing anything special there) and more importantly, current implementation has the side-effect of expanding relative file names (not just symlinks, any relative files) in metadata, should I change it to keep the old behavior or adjust affected tests and leave it that way? Many thanks for pointers and opinion.

**Comment by rgerhards** [comment]
> I have two questions now. First, how can I determine which part of my test causes the "permission denied" failures (as far as I am aware I am not doing anything special there) 

Let me check. I am still working on CI (you probably have seen many new buildbot slaves plus Fedora28). This could have caused it. I'll have a look at the whole thing and let you know if something needs attention. From a first glimpse, it looks like we have some path expansion messages that do not belong there. But I may be wrong here. Will have a deeper look when I am through with the current round of CI fixing, as it may affect what you see (so it does not make sense to look immediately).

> and more importantly, current implementation has the side-effect of expanding relative file names (not just symlinks, any relative files) in metadata, should I change it to keep the old behavior or adjust affected tests and leave it that way? 

Good question. @davidelang what do you think? I would tend to say it should stay as it was, but for symlinks it probably is questionable.

**Comment by davidelang** [comment]
On Wed, 18 Jul 2018, Rainer Gerhards wrote:

> Good question. @davidelang what do you think? I would tend to say it should 
> stay as it was, but for symlinks it probably is questionable.

I think we need to keep things as they are, make the expanded version be 
additional metadata

for kubernetes, we depend on three layers of symlinks, and the name of the first 
symlink contains critical information that doesn't exist in the other two 
layers. Ideally, each layer should be documented in the metadata.

David Lang

**Comment by richm** [comment]
For the kubernetes case, we would need to use a function other than `realpath(3)` to get the target.  We would have to make one or more calls to `readlink(2)` to follow an entire chain of links to get every intermediate link.

**Comment by richm** [comment]
The way it works currently (i.e. without this patch/PR) is that the name of the symlink is used for `$!metadata!filename`.  In the kubernetes case, if I have
```
input(type="imfile" file="/var/log/containers/*.log" tag="kubernetes" addmetadata="on")
```
Then the record contains
```
$!:{ "metadata": { "filename": "\/var\/log\/containers\/logging-kibana-1-ctwqh_openshift-logging_kibana-1ce792f7c50f2b2e38d56050f32e6c527589515e299ee2670a099a2d08361f79.log", "fileoffset": "5955682" } }
```
instead of some target file under `/var/log/pods/` or `/var/lib/docker/containers/*`.  This behavior is necessary for mmkubernetes to work.

**Comment by davidelang** [comment]
On Wed, 18 Jul 2018, Richard Megginson wrote:

> For the kubernetes case, we would need to use a function other than 
> `realpath(3)` to get the target.  We would have to make one or more calls to 
> `readlink(2)` to follow an entire chain of links to get every intermediate 
> link.

From a metadata point of view, we can do three things.

1. show the filename of the first link (what rsyslog config is told to look at)

2. show the path to the actual file

3. show the filename of the first link, and then in separate fields show each 
intemediate link and the final path to the actual file

4. show the filename of the first link, then in a separate field show the path 
to the actual file


kubernetes will break if the filename for the first link isn't available.

I can see people wanting the other information.

so in order of preference, I would say 3, 4, 1, and far behind 2.

to avoid breaking existing configs, we need to leave the filename metadata 
showing what it does today (which would rule out 2).

adding the metadata for the final path in a separate variable (4) would be an 
improvement over what we do today. But adding a third variable that's an array 
of any intermediate links would be a nice-to-have.

David Lang

**Comment by jvymazal** [comment]
> I would like to do this in two steps.
> 
> First step - this current PR should be changed to do 1. show the filename of the first link (what rsyslog config is told to look at) - this will preserve the current behavior and not introduce any regressions

Understood, I will edit the PR to carry over the link name to the target. However this should definetely be documented as my first understanding after reading https://github.com/rsyslog/rsyslog-doc/blob/master/source/configuration/modules/imfile.rst#metadata was that filename should behave like 2. which is why I originally written it that way (should have tested that the code actually does that).

**Comment by rgerhards** [comment]
> Understood, I will edit the PR to carry over the link name to the target. However this should definetely be documented as my first understanding after reading https://github.com/rsyslog/rsyslog-doc/blob/master/source/configuration/modules/imfile.rst#metadata was that filename should behave like 2.

PR would be appreciated ;-)

**Comment by jvymazal** [comment]
However, if preserving 

> 1. show the filename of the first link (what rsyslog config is told to look at)

this raises a question of what to list there in case there are more symlinks pointing to the same file. For now I will go with using the first symlink I encounter and ignoring the rest. @richm @davidelang any comments to this?

**Comment by rgerhards** [comment]
> this raises a question of what to list there in case there are more symlinks pointing to the same file. For now I will go with using the first symlink I encounter and ignoring the rest.

As the gouverning rule I would say "it should behave exactly like it did before the patch". So whatever it currently does is the right thing to do.

**Comment by rgerhards** [comment]
side-note: it looks like the CI failures are valid regressions, so something to check out.

I do not know what the permission issue is. Maybe add ```set -x ; set -v``` on top of the test?

**Comment by rgerhards** [comment]
If you have problems to reproduce the issues locally, you can use the docker container CI uses. It's easy to do via ```devtools/devcontainer``` - see that file for the env vars to set.

**Comment by richm** [comment]
imfile.c has a potential uninitialized variable - this fixes it:
```
diff --git a/plugins/imfile/imfile.c b/plugins/imfile/imfile.c
index 6c5ebb65a..8d4a7a6d9 100644
--- a/plugins/imfile/imfile.c
+++ b/plugins/imfile/imfile.c
@@ -810,7 +810,6 @@ poll_tree(fs_edge_t *const chld)
 {
        struct stat fileInfo;
        glob_t files;
-       int issymlink;
        DBGPRINTF("poll_tree: chld %p, name '%s', path: %s\n", chld, chld->name, chld->path);
        detect_updates(chld);
        const int ret = glob((char*)chld->path, runModConf->sortFiles|GLOB_BRACE, NULL, &files);
@@ -818,6 +817,7 @@ poll_tree(fs_edge_t *const chld)
        if(ret == 0) {
                DBGPRINTF("poll_tree: processing %d files\n", (int) files.gl_pathc);
                for(unsigned i = 0 ; i < files.gl_pathc ; i++) {
+                       int issymlink = 0;
                        if(glbl.GetGlobalInputTermState() != 0) {
                                goto done;
                        }
```

**Comment by richm** [comment]
I think the tests will pass if you add `imfile-symlink.sh` to `EXTRA_DIST`in tests/Makefile.am

**Comment by davidelang** [comment]
On Thu, 19 Jul 2018, jvymazal wrote:

> However, if preserving
>
>> 1. show the filename of the first link (what rsyslog config is told to look at)
>
> this raises a question of what to list there in case there are more symlinks pointing to the same file. For now I will go with using the first symlink I encounter and ignoring the rest. @richm @davidelang any comments to this?

for 1, pretend that you know nothing about symlinks, you have a filename that 
you open (either from the config or from matching a pattern), that is what you 
open, and you don't care if it's the only link to a file, an additional hard 
link, a symlink, a named pipe, or anything else. You get back a filehandle that 
you can issue reads on.

That name is what should be reported in the metadata today. As far as the 
application is concerned, that's the filename.

Now, in the case of symlinks, there is additional knowledge that we can gather, 
what the final name is, and what the hops are to get there.

In no case should we be searching to find if we have multiple links to the same 
file.

unless, are you saying that we follow the links to be able to detect rotation of 
the file pointed at by the symlink, and because we do so, we may discover that 
we have multiple filehandles open to the same file?

In this case, my tendency would be to log a warning at file open time, and then 
happily duplicate the logs by reading them through each original filename. After 
all, that's what the admin told us to do. It may seem strange, but we had 
someone just this week who was doing this deliberately because they wanted to 
send the logs to different destinations with different facilities on the log 
message, so he created symlinks and had multiple imfile input() sections?

David Lang

**Comment by jvymazal** [comment]
> for 1, pretend that you know nothing about symlinks, you have a filename that you open (either from the config or from matching a pattern), that is what you open, and you don't care if it's the only link to a file, an additional hard link, a symlink, a named pipe, or anything else. You get back a filehandle that you can issue reads on. That name is what should be reported in the metadata today. As far as the application is concerned, that's the filename. 
This is somehow what actual state of the PR is with the last edits.

> unless, are you saying that we follow the links to be able to detect rotation of the file pointed at by the symlink, and because we do so, we may discover that we have multiple filehandles open to the same file? 
Unless making some major code changes there will never be more open file handles to one file, this is resolved by inotify code, not us.
What is happening currently is that if you configure more symlinks to same file, when creating data for 2nd+ symlink I find that there is already record in the table. To be consistent with original behavior I am skipping the processing of such files now. Only difference is that before you got wd error and now you get specific log notice stating which superfluous links to which file we found.

**Comment by rgerhards** [comment]
> Unless making some major code changes there will never be more open file handles to one file, this is resolved by inotify code, not us.

we do rely on inotify only for notifications. As they are not reliably reported, we still iterate over our data structures to see if a new file popped up. I guess that means your argument is wrong.

**Comment by rgerhards** [comment]
> What is happening currently is that if you configure more symlinks to same file, when creating data for 2nd+ symlink I find that there is already record in the table. To be consistent with original behavior I am skipping the processing of such files now. Only difference is that before you got wd error and now you get specific log notice stating which superfluous links to which file we found.

sorry, I should have taken a bit more time ;-)

I stil think the existing code processed the file, it just did not manage to setup an inotify watch. Is that right? If so, I think it has implications on what we need to preserve.

**Comment by rgerhards** [comment]
also, we must ensure everything works when no inotify is available, e.g. BSD or Solaris. Do we need additional tests to ensure this is the case?

**Comment by jvymazal** [comment]
> also, we must ensure everything works when no inotify is available, e.g. BSD or Solaris. Do we need additional tests to ensure this is the case?

Apart from the watches the code should work in polling mode as well, so targets will be added to active lists and polled.

**Comment by rgerhards** [comment]
> Apart from the watches the code should work in polling mode as well, so targets will be added to active lists and polled.

```should``` or ```does```? ;-)

**Comment by rgerhards** [comment]
In other words: I think we need a test case where two symlinks point to the same file and we must check that in the end result the file is processed twice. @davidelang did I get this right?

**Comment by jvymazal** [comment]
> we must check that in the end result the file is processed twice.

OK, I think I get it now, @davidelang in example you posted I believe polling mode was used, right?  In that case it needs one more tweak to handle this situation. I was not aware that there was symlink support for the polling mode as well.

**Comment by rgerhards** [comment]
The refactored imfile (8.35+?) has unified inotify/FEN/polling. even in inotify there is some polling mode involved, as the inotify API is not reliable (we had many bug reports that rooted in that). So inotify/FEN are primarily used for hinting us in the new code.

**Comment by rgerhards** [comment]
For clarity, I should add that the exiting code does not care about symlinks at all. So whatever currently happens with them, is not by design but by "accident". Nevertheless, people probably count on the existing behavior, so we cannot break it by default. Even though the current handling was never explicitly designed, but it is in place for many years now.

**Comment by jvymazal** [comment]
Understood, I will add a case for this so multiple links are still processed if configured that way.

**Comment by jvymazal** [comment]
@rgerhards one more question, after adding new test to distcheck section I get an error that there are some files left after dist clean, where should I add those to fix it please?

**Comment by rgerhards** [comment]
@jvymazal  thx, sorry for not being specific about the refactoring in the immediately. If in doubt, I think you can run the test against the codebase without this patch. Whatever it does, is expected behavior (I admit I have never looked at that case, so I am really not sure what happens). If the current code really makes strange things, we possibly need to reconsider if there is a bug in existing code, but from @davidelang I got the impression that it works in a defined way.

I would also like to thank you for being so persistent with what looked like a simple patch but turns into something much more elaborate. I know that feeling...

**Comment by rgerhards** [comment]
> @rgerhards one more question, after adding new test to distcheck section I get an error that there are some files left after dist clean, where should I add those to fix it please?

If they are something created by the test, they should be removed by then init/exit here https://github.com/rsyslog/rsyslog/blob/master/tests/diag.sh#L136

We also remove by init to ensure that the next test gets a clean environment (error-exit intentionally does not clean up).

File: plugins/imfile/imfile.c [hunk 5]
```diff
@@ -746,13 +758,44 @@ poll_active_files(fs_edge_t *const edge)
...
+		FINALIZE;
+	}
+	act_obj_add(chld, target, is_file, fileInfo.st_ino, 0, symlink);
+
+finalize_it:
+	free(target);
+	RETiRet;
+}
 
-static void ATTR_NONNULL() poll_tree(fs_edge_t *const chld);
```
Comment (rgerhards) [inline plugins/imfile/imfile.c:750]: I think I remember this line was required by some analyzer - but I may be wrong. CI will hopefully show,

File: plugins/imfile/imfile.c [hunk 6]
```diff
@@ -1384,6 +1452,9 @@ pollFile(act_obj_t *const act)
 {
 	cstr_t *pCStr = NULL;
 	DEFiRet;
+	if (act->is_symlink) {
+		RETiRet;    /* no reason to poll symlink file */
```
Comment (rgerhards) [inline plugins/imfile/imfile.c]: This should be FINALIZE (RETiRet must only be used at end of function, in some settings it can generate extra code).

File: plugins/imfile/imfile.c (hunk 5 above)
Comment (jvymazal) [reply plugins/imfile/imfile.c:750]: I think that last run almost passed (minus the extra file left but actual tests were green)

**Comment by davidelang** [comment]
On Fri, 20 Jul 2018, Rainer Gerhards wrote:

> In other words: I think we need a test case where two symlinks point to the 
> same file and we must check that in the end result the file is processed 
> twice. @davidelang did I get this right?

Yes.

David Lang

**Comment by jvymazal** [comment]
OK, @rgerhards, changed the implementation to check for both name and source before skipping act_obj_add() and also added another test case for this.

**Comment by jvymazal** [comment]
Fine, I will push it, I just have to solve last possible bug, as scenario was found where everything looks like `if (S_ISLNK(fileInfo.st_mode))` on symlink file is reporting false, will have to investigate it so we do not have to merge this twice

**Comment by jvymazal** [comment]
Ok, found there there is a difference between `stat()` and `lstat()` results.. now it will hopefully work as intended, also edited the newly added test as per your recent testbench commit, @rgerhards if it goes all green we should be fine to merge.

**Comment by rgerhards** [comment]
thx a lot!

**Comment by rgerhards** [comment]
side-note, just FYI: there is still some instability in CI. I am working on that. If something turns red, I'll definitely have a look. It may be unrelated.

File: tests/imfile-symlink-multi.sh [hunk 7]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+# This is part of the rsyslog testbench, licensed under GPLv3
```
Comment (rgerhards) [inline tests/imfile-symlink-multi.sh]: sry, a final nit :-( could you license under ASL 2.0 if you have no concerns with that (if there are, keep as is!). We try to move toward ASL 2.0, but some old tests still bear GPLv3.

File: tests/imfile-symlink.sh [hunk 8]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+# This is part of the rsyslog testbench, licensed under GPLv3
```
Comment (rgerhards) [inline tests/imfile-symlink.sh]: ASL 2.0 preferred

**Comment by jvymazal** [comment]
> I did not find the test mentioned here #2749 (comment) - I would also assume that this requires a new config parameter.

The imfile-symlink-multi tests just that (it points 10 symlinks to one target file and then checks there are 10 messages in out.log
I do not see why we should need any new parameter for that.

**Comment by rgerhards** [comment]
> The imfile-symlink-multi tests just that (it points 10 symlinks to one target file and then checks there are 10 messages in out.log
> I do not see why we should need any new parameter for that.

maybe I am doing too many things in parallel, but... By default, imfile should behave like before the patch. So it must report the unresolved symlink name. Does the test do that? And if it does, how does imfile know when it has to report the resolved symlink?

**Comment by rgerhards** [comment]
> By default, imfile should behave like before the patch. So it must report the unresolved symlink name. Does the test do that?

I withdraw this comment - the test does indeed do that.

Now let me tripple-check the other question.

**Comment by jvymazal** [comment]
@rgerhards regarding the now-failing static analyzer, would you mind taking a look at the report? I have tried to go trough it several times but each time I came to conclusion that it is a false positive. I actually changed the very lines it complains about after discovering and fixing segfault in them. Thanks

**Comment by rgerhards** [comment]
> regarding the now-failing static analyzer, would you mind taking a look at the report?

will have a look at the moment, got distracted yesterday.

To my other question and without digging deeper: how is the new behavior of symlink processing controlled? I did not see any config switch.

**Comment by jvymazal** [comment]
> To my other question and without digging deeper: how is the new behavior of symlink processing controlled? I did not see any config switch.

Well, I did not add any :-)
I tried to keep it in line with original behavior with the addition the code is now actually aware of symlinks and has separate watches and objects for them AND for their targets. I do not see why we should need any new switch for that. If you think there are some cases where current state of PR would break something please bring it up.

**Comment by rgerhards** [comment]
> I tried to keep it in line with original behavior with the addition the code is now actually aware of symlinks and has separate watches and objects for them AND for their targets. I do not see why we should need any new switch for that. If you think there are some cases where current state of PR would break something please bring it up.

mmmhhhh... if there is absolutely no change of behavior vs. previous code, I think we could leave all as is ;-) That's just the philosophical PoV, of course. I'll see that I do some tests if all is kept active and have a deeper look at the new code. 

@davidelang @richm would also appreciate your feedback, as I did not yet have symlinks on my mind.

File: plugins/imfile/imfile.c [hunk 9]
```diff
@@ -702,27 +714,24 @@ detect_updates(fs_edge_t *const edge)
...
-	for(act = edge->active ; act != NULL ; ) {
+	for(act = edge->active ; act != NULL ; act = act->next) {
 		DBGPRINTF("detect_updates checking active obj '%s'\n", act->name);
-		const int r = stat(act->name, &fileInfo);
+		const int r = lstat(act->name, &fileInfo);
 		if(r == -1) { /* object gone away? */
 			DBGPRINTF("object gone away, unlinking: '%s'\n", act->name);
-			act_obj_t *toDel = act;
-			act = act->next;
-			DBGPRINTF("new next act %p\n", act);
```
Comment (rgerhards) [inline plugins/imfile/imfile.c:713]: 1: I guess the static analyzer is somewhat confused, but isn't the removal of this line (plus the next I flag) the root problem?

File: plugins/imfile/imfile.c [hunk 10]
```diff
@@ -702,27 +714,24 @@ detect_updates(fs_edge_t *const edge)
...
-			act_obj_unlink(toDel);
-			continue;
+			act_obj_unlink(act);
```
Comment (rgerhards) [inline plugins/imfile/imfile.c:727]: 2: now act is deleted (freed)

File: plugins/imfile/imfile.c [hunk 11]
```diff
@@ -702,27 +714,24 @@ detect_updates(fs_edge_t *const edge)
 {
 	act_obj_t *act;
 	struct stat fileInfo;
+	int restart = 0;
 
-	for(act = edge->active ; act != NULL ; ) {
+	for(act = edge->active ; act != NULL ; act = act->next) {
```
Comment (rgerhards) [inline plugins/imfile/imfile.c:722]: 3: act in act->next has been freed in line 724 (see comment below).

File: plugins/imfile/imfile.c (hunk 9 above)
Comment (rgerhards) [reply plugins/imfile/imfile.c:713]: sry, previous line 711 I meant

File: plugins/imfile/imfile.c (hunk 11 above)
Comment (jvymazal) [reply plugins/imfile/imfile.c:722]: Yes, however, if I unlink it I break the cycle and the processing begins again from the start.

Comment (rgerhards) [reply plugins/imfile/imfile.c:722]: but you actually unlink it in 724?

Comment (jvymazal) [reply plugins/imfile/imfile.c:722]: The problem I see when analyzing the result seems to be that clang seems to ignore [this line ](https://github.com/rsyslog/rsyslog/pull/2749/files#diff-b563a9e8d7f28e380e7fd4dc4467d3e8R969) line which guards all 3 cases it reports.

**Comment by rgerhards** [comment]
2018-07-25 10:18 GMT+02:00 jvymazal <notifications@github.com>:

> *@jvymazal* commented on this pull request.
> ------------------------------
>
> In plugins/imfile/imfile.c
> <https://github.com/rsyslog/rsyslog/pull/2749#discussion_r205020166>:
>
> >
> -	for(act = edge->active ; act != NULL ; ) {
> +	for(act = edge->active ; act != NULL ; act = act->next) {
>
> The problem I see when analyzing the result seems to be that clang seems
> to ignore this line
> <https://github.com/rsyslog/rsyslog/pull/2749/files#diff-b563a9e8d7f28e380e7fd4dc4467d3e8R969>
> line which guards all 3 cases it reports.
>
>
> Have your read *all* my comments? clang is wrong, but it is confused by a
problem that it does not properly report. The problem itself is present.

File: plugins/imfile/imfile.c (hunk 11 above)
Comment (rgerhards) [reply plugins/imfile/imfile.c:722]: the problem is not act->name but act itself.

Comment (rgerhards) [reply plugins/imfile/imfile.c:722]: act->next leads to deref of act, which has been freed in line 724.

Comment (jvymazal) [reply plugins/imfile/imfile.c:722]: Yes, but if I free it I never get here again, see comments on below lines.

File: plugins/imfile/imfile.c [hunk 12]
```diff
@@ -702,27 +714,24 @@ detect_updates(fs_edge_t *const edge)
...
 		if(r == -1) { /* object gone away? */
 			DBGPRINTF("object gone away, unlinking: '%s'\n", act->name);
-			act_obj_t *toDel = act;
-			act = act->next;
-			DBGPRINTF("new next act %p\n", act);
-			act_obj_unlink(toDel);
-			continue;
+			act_obj_unlink(act);
+			restart = 1;
+			break;
```
Comment (jvymazal) [inline plugins/imfile/imfile.c:729]: I had to change a flow now because unlink call can do more thinks due to additional target removals, so if I unlink anything flom list I abort completely and go trough whole function again so I have ensured coherent state of the list.

Comment (rgerhards) [reply plugins/imfile/imfile.c:729]: ok, let me first try to update to clang-6.0 analyzer. Maybe it solves the situation. This also sounds Iike I do need to go more in-depth over the PR. I'll ping you when clang-6 is ready, not sure I will mange this today.

**Comment by rgerhards** [comment]
@jvymazal FYI: I gave changing to clang-6.0 a try, but it found some yet-undetected issues: https://github.com/rsyslog/rsyslog/pull/2875 -- will need to work on them before I can activate.

**Comment by davidelang** [comment]
I see no reason for having a config item for this.

People currently using metadata will keep working, there will just be some extra 
variables that they don't do anything with.

David Lang

**Comment by rgerhards** [comment]
> I see no reason for having a config item for this.
> 
> People currently using metadata will keep working, there will just be some extra
> variables that they don't do anything with.

thx, this clears my concern

@jvymazal can you please rebase, looks like my patch to switch the clang 6 affected your commit. That means there is a good chance it could go through with the next run.

**Comment by jvymazal** [comment]
done

**Comment by jvymazal** [comment]
@rgerhards the only current failure seems unrelated (only kafka test), otherwise the static analyzer looks resolved, thanks!

**Comment by jvymazal** [comment]
@rgerhards I did another rebase as there were some errors from distclean, hope it clears them, do you think that you can manage the code review till 8.37 ?

**Comment by rgerhards** [comment]
Sry, I just did the final merge :-( I cannot do late merges this time as I am out of office next week and so all release testing needs to be done today.

**Comment by rgerhards** [comment]
@jvymazal  sorry I didn't make the 8.37 milestone, but at least now it is merged! Thanks for your persistence!

**Comment by rgerhards** [comment]
@jvymazal thinking about it, I think the doc should also be amended, at least re the new metadata items. Could you have a look at it?

**Comment by jvymazal** [comment]
@rgerhards there are actually no new metadata items that would need documentation, and the behavior regarding original metadata was intentionally left unchanged (with rsyslog/rsyslog-doc/pull/707 actually explaining the behavior), see discussion above. 
What is new is possibility to easily add new metadata field which would contain the actual, absolute name of the file, but that was not done in scope of this.

**Comment by rgerhards** [comment]
@jvymazal  sorry, I am puzzled again. Maybe I should have looked at that aspect in depth.

@davidelang said

> there will just be some extra variables that they don't do anything with.

I had assumed that this PR adds additional metadata, which a user can use but also can ignore. If the PR does not do that, do we expect a follow-up PR to use the new functionality? Otherwise I am back to the question why the change was made at all, as everything then seemed to work as before, just more complicated. I am sure I overlooked something...

File: plugins/imfile/imfile.c [hunk 13]
```diff
@@ -1162,7 +1222,11 @@ enqLine(act_obj_t *const act,
 	msgSetPRI(pMsg, inst->iFacility | inst->iSeverity);
 	MsgSetRuleset(pMsg, inst->pBindRuleset);
 	if(inst->addMetadata) {
-		metadata_values[0] = (const uchar*)act->name;
+		if (act->source_name) {
```
Comment (rgerhards) [inline plugins/imfile/imfile.c:1225]: mhhh... Isn't that changing the existing behavior?

Comment (jvymazal) [reply plugins/imfile/imfile.c:1225]: No, it is not, this is ensuring that what was changed in the internal structures does *not* affect what user has in metadata.

**Comment by jvymazal** [comment]
> Otherwise I am back to the question why the change was made at all, as everything then seemed to work as before, just more complicated.

The original problem was that the structures before were not awware of the symlonks (we relied on filesystem to somehow make it work) which caused that imfile stopped watching rotated files which were loaded via symlink, this should not happen with the new code.

**Comment by rgerhards** [comment]
> The original problem was that the structures before were not aware of the symlinks (we relied on file-system to somehow make it work) which caused that imfile stopped watching rotated files which were loaded via symlink, this should not happen with the new code.

ahhhh! this should probably have went into the commit message, I guess a lot of things would have run smoother ;-) I now see there is an issue ref in the second message, but these scroll away so soon...

The past couple of weeks (?) I was under the impression this is a functional enhancement.

**Comment by jvymazal** [comment]
> The past couple of weeks (?) I was under the impression this is a functional enhancement.

Well, that depend on how you define bugs and features ;-) , as such rsyslog never had any explicit symlink support so the issue could be regarded as feature request rather than bug...

**Comment by rgerhards** [comment]
> Well, that depend on how you define bugs and features ;-) , as such rsyslog never had any explicit symlink support so the issue could be regarded as feature request rather than bug...

ack - I could also have re-read things more often, but I admit I usually look at the commit message and it was missing the fact that the primary target was to enable something that worked strange for other folks. I should have noticed, but I did not - sry.

My friendly suggestion here is just to make the commit message cristal clear, a link to an existing issue is also usually a life saver (at least for me).

That said, all is well, but I wanted to tell what happened on my side ;-)

**Comment by jvymazal** [comment]
> My friendly suggestion here is just to make the commit message cristal clear, a link to an existing issue is also usually a life saver (at least for me).

Ok, will try to remember that for future PRs

**Comment by davidelang** [comment]
I'll also say that it's common for people to be watching long-running issues 
like this in e-mail (and/or just look at the last couple of messages in a long 
thread), so don't assume that people have the entire context from the beginning 
in their head when looking at a message, try to include enough context (or at 
least a pointer to where to refresh themselves) rather than assuming they will 
go through the entire thread.

David Lang
